Allow negative bit-shift counts

eschnett opened the pull request:
This patch concerts the bit shift operators << >> >>>.
- Convert signed bit-shift counts explicitly to unsigned, reporting InexactErrors for negative counts.
- Introduce versions that take unsigned arguments, and which thus do not require any tests.
- Introduce new functions `bitshift` and `unsigned_bitshift` that (similar to Matlab) shift left or right for positive and negative counts, respectively.
- Add documentation.
- Add tests.

Closes #14516.

tkelman commented:
`unsigned_bitshift` is not a name we should export. What's the simplest description of the difference between `x >> -n` and `x >>> -n` here?

eschnett commented:
Should we even have `unsigned_bitshift`? How commonly is `>>>` used? This function is not necessary; one can always convert to the respective unsigned type instead.

In machine language, this would be called a "logical shift".

simonbyrne commented:
We could use `<>` or `<<>` as operators?

tkelman commented:
Leaving it out sounds appropriate.

simonbyrne commented:
`>>>` appears 109 times in Base, `>>` 372.

eschnett commented:
Removed `unsigned_bitshift`, rebased.

tkelman commented:
LGTM. @StefanKarpinski what do you think?

tkelman commented:
IIUC this is breaking, so it could use a mention in NEWS.md.

StefanKarpinski commented:
I'm not sure. I'm somewhat more inclined to define `<<`, `>>` and `>>>` to work "correctly" with negative shift values – and not introduce the `bitshift` function. The generated code is kind of long, but potentially throwing an exception is also a total performance disaster, so anyone who's not using shift values that the compiler can reason about who cares about performance would need to do additional work anyway.

StefanKarpinski commented:
@JeffBezanson, thoughts?

eschnett commented:
Ping

eschnett commented:
@StefanKarpinski: It seems @JeffBezanson doesn't have input on this. What's your verdict?

StefanKarpinski commented:
Rather than introducing a new `bitshift` function, let's make negative shifts work for existing shift ops:

```jl
a  << -b == a >> b
a  >> -b == a << b
a >>> -b == a << b
```

I suspect that in this case, using a branch will be better than using a select instruction (`ifelse`) since the branch will tend to be very predictable. Otherwise keep the rest of this change – i.e. making the unsigned version fundamental, so that you can avoid the branch entirely by using an unsigned shift value.

@eschnett, do you mind modifying this PR this way (or making a new one)?

simonbyrne commented:
> Rather than introducing a new bitshift function, let's make negative shifts work for existing shift ops:

+1

carnaval commented:
as long as constant shifts get folded correctly I think that's fine

eschnett commented:
I think might be worth a benchmark.

simonbyrne commented on base/gmp.jl:
```diff
@@ -346,23 +346,27 @@ for (fJ, fC) in ((:-, :neg), (:~, :com))
 
-function <<(x::BigInt, c::Int)
-    c < 0 && throw(DomainError())
+function <<(x::BigInt, c::UInt)
```
Shouldn't this be `Culong`?

eschnett replied:
Making this `Culong` would force every caller to explicitly convert to `Culong`.

simonbyrne replied:
Ah, I misread it (I thought it was `Cuint`).

What if `Culong != UInt` (as I think it is on Windows 64-bit): won't the implicit conversion throw an error for `typemax(UInt)`?

eschnett replied:
Possibly; I don't know whether `ccall` checks. 64-bit Linux definitely has this problem. That problem was there before, though.

I don't know a good solution -- shifting a `BigInt` left by a large number should work "in principle", but will in practice run out of memory. So an error message is probably the best we can do anyway.

simonbyrne replied:
That's probably true.

simonbyrne commented on base/int.jl:
```diff
@@ -162,6 +157,26 @@ trailing_ones(x::Integer) = trailing_zeros(~x)
+<<(x::BitInteger, y::BitSigned) =
+    0 <= y ? x << unsigned(y) : x >> unsigned(-y)
+>>>(x::BitInteger, y::BitSigned) =
+    0 <= y ? x >>> unsigned(y) : x << unsigned(-y)
```
Can we make these more generic so we don't need the extra methods in `gmp.jl`?

eschnett replied:
Good idea.

eschnett commented:
Are there any further comments or requests? I'm ready to squash and merge.

eschnett committed:
Support negative bit-shift counts for <<, >>, and >>>

- Introduce methods that take unsigned shift counts
- Handle signed shift counts by dispatching to the respective unsigned shift counts, shifting in the opposite direction for negative counts
- Update documentation
- Add tests

Closes #14516.

eschnett commented:
Rebased and squashed, Travis and Appveyor are green. Ready to go?

tkelman commented on base/operators.jl:
```diff
@@ -126,12 +126,16 @@ end
-where `n >= 0`, filling with `0`s. This is equivalent to `x * 2^n`.
+Left bit shift operator, `x << n`. For `n >= 0`, the result is `x` shifted left
+by `n` bits, filling with `0`s. This is equivalent to `x * 2^n`. For `n < 0`,
+this is equivalent to `x >> -n`.
```
needs a `make docs` to update the rst contents

eschnett replied:
Done. Anything else I missed?

tkelman replied:
I think we're good.

tkelman commented:
This seems to have introduced several ambiguity warnings during bootstrap.

eschnett commented:
I will check.

eschnett commented:
This is now PR #15368.

tkelman commented:
> IIUC this is breaking, so it could use a mention in NEWS.md.

This was never addressed here, was it?

eschnett commented:
Good catch.

eschnett commented:
@tkelman In what sense in this breaking? The new functionality is a strict superset of the old one. Do you refer to how one now defines bit-shift operations for user-defined types? Or do you refer to the performance difference between signed and unsigned shift counts?

tkelman commented:
None of those - weren't we previously doing the Perl behavior for https://github.com/JuliaLang/julia/issues/14516#issuecomment-168229923, so after this PR negative shifts now give different results?

tkelman commented:
None of those - weren't we previously doing the Perl behavior for https://github.com/JuliaLang/julia/issues/14516#issuecomment-168229923, so after this PR negative shifts now give different results?

eschnett commented:
Very good; yes, I understand now. I'll add the `NEW.md` entry.

eschnett commented:
See ##18188.
